yellow-paper/docs/contracts/index.md

```diff
@@ -12,6 +13,138 @@ The purpose of the L1 contracts are simple:
+    available: Map[Fr => bool]
+    mut publish(effects: TxEffect[]): Fr
+    is_available(txs_hash: Fr): bool
+}
+
+class Inbox {
+    consume(entries: Fr[])
+}
+
+class Outbox {
+    insert(entries: Fr[])
+}
+
+class Verifier {
+    verify(proof: Proof, header: Header, archive: Fr): bool
+}
+
+class StateTransitioner {
+    archive: Snapshot
+    process(header: Header, archive: Fr, txs_hash: Fr, l1_to_l2_msgs: Fr[], l2_to_l1_msgs: Fr[], proof: Proof)
+}
+StateTransitioner --> AvailabilityOracle: is_available()
+StateTransitioner --> Inbox: consume()
+StateTransitioner --> Outbox: insert()
+StateTransitioner --> Verifier: verify()
+```
+
+
+### State transitioner
+The state transitioner is the heart of the validating light node for the L2. The contract keeps track of the current state of the L2 and progresses this state when a valid L2 block is received. It also facilitates cross-chain communication (communication between the L1 inbox and outbox contracts).
+
+```python
+class StateTransitioner:
+
+    def __init__(
+        self, 
+        verifier: Verifier,
+        registry: Registry,
+        version: Fr, 
+        archive: Snapshot
+    ):
+        self.verifier = verifier
+        self.registry = registry
+        self.version = version
+        self.archive = archive
+        self.last_block_ts = block.timestamp
+        self.block_number = 0
+
+    def process(
+        self,
+        header: Header, 
+        archive: Fr, 
+        txs_hash: Fr, 
+        l1_to_l2_msgs: Fr[], 
+        l2_to_l1_msgs: Fr[], 
+        proof: Proof
+    ):
+        assert self.body_available(header, txs_hash, l1_to_l2_msgs, l2_to_l1_msgs)
+        assert self.validate_header(header)
+        assert self.verifier.verify(proof, header, archive)
+        assert self.registry.inbox.consume(l1_to_l2_msgs)
+        assert self.registry.outbox.insert(l2_to_l1_msgs)
+        self.archive = archive
+
+    def body_available(
+        self,
+        content_hash: Fr, 
+        txs_hash: Fr, 
+        l1_to_l2_msgs: Fr[], 
+        l2_to_l1_msgs: Fr[]
+    ) -> bool:
+        assert self.registry.availability_oracle.is_available(txs_hash)
+        in_hash = SHA256(pad(l1_to_l2_msgs))
+        out_hash = MerkleTree(pad(l2_to_l1_msgs), SHA256)
+        return content_hash == SHA256(txs_hash, out_hash, in_hash)
+    )
+
+    def validate_header(
+        self,
+        header: Header
+    ) -> bool:
+        assert header.global_variables.block_number = self.block_number + 1
+        assert header.global_variables.chain_id == block.chain_id
+        assert header.global_variables.version == self.version 
+        assert header.global_variables.timestamp < block.timestamp
+        assert header.global_variables.timestamp > self.last_block_ts
+        assert header.archive == self.archive
+
+        return True
+```
+
+
+### Availability Oracle
+The state transitioner should be connected to an oracle which addresses the availability condition. 
```

Comment:
What's "the availability condition"?

Reply:
That the data must be published and available for later consumption.